### ISSUE TITLE
IntelliJ Support

### ISSUE THREAD

**Issue by dwijnand** [coursier/sbt-launcher#26]
I've been using this quite successfully.

I had to workaround #3 by adding adding `classpathTypes += "maven-plugin"` and I still need to go back and verify #21 is fixed (for me), but it's been great.

However, IntelliJ IDEA uses its own, bundled sbt-launcher.  I could override that setting and link it to, for example, `~/.sbt/launchers/coursier_1.2.5/sbt-launch.jar` but what happens when I try and import an sbt 0.13 projects?  Or maybe a project on a slightly older version of sbt 1?  Also I'd probably want to keep using the latest (and greatest) version of coursier's sbt-launcher, so I'd have to update the jar to use (in one way or another).

Would it be possible to do some coordination and collaboration with @jastice of JetBrains to bring coursier's sbt-launcher support to IntelliJ IDEA?  Maybe in some experimental capacity.  I'm happy to help test the nightlies of IntelliJ Scala.

**Comment by jastice** [comment]
What *does* happen if you change the setting in IntelliJ? Wouldn't the this only change the behavior of the launcher itself?

I'm happy to look into this next release cycle (April/May)

**Comment by dwijnand** [comment]
No, using the coursier sbt-launcher makes all parts of sbt use coursier, for both the build dependencies and the project dependencies.

**Comment by jastice** [comment]
That's neat. From IntelliJ-side we could have some kind of option to use this launcher when a certain required sbt version is configured, for example.

**Comment by dwijnand** [comment]
Works for me!

Perhaps @alexarchambault can give provide that info: what versions of sbt would coursier's sbt-launcher work for?

**Comment by alexarchambault** [comment]
I ought to work back to sbt 0.13.8, the first sbt 0.13 version sbt-coursier was compatible with. (And seems to in a few manual tests, and in the [automatically tested project](https://github.com/coursier/sbt-launcher/blob/8d53da9c8f1f5fca73957075e88d4161dec3741c/sbt-launcher/src/test/scala/coursier/sbtlauncher/SbtPre12Tests.scala#L13-L29).)

Never tested it with sbt versions prior to that one. The launcher may need to be modified so that it doesn't try to inject sbt-coursier upfront with those.

I don't know in what extent the mainline sbt launcher changed during 0.13.

**Comment by alexarchambault** [comment]
Intellij seems to be fine using the JAR of the coursier sbt-launcher as sbt JAR. (Although when I try it, the setting seems to revert back to bundled launcher on its own sometimes, I'm not sure why.)

**Comment by alexarchambault** [comment]
@jastice Is there a way to have the launcher, or an sbt plugin, report some progress directly to IntelliJ? It would be nice if the launcher or sbt-coursier could display progress bars in IntelliJ, like those we get when importing a Maven project.

**Comment by jastice** [comment]
There's always *some* way, but I'd have to look into how practical it is. Communicaton with sbt is still only by parsing shell output. These days I like to move everything on BSP ;) which has a progress reporting API that I already support. But it would be a bit of work to make sbt support it as well.

**Comment by dwijnand** [comment]
So I tried out using `~/.sbt/launchers/coursier_1.2.5/sbt-launch.jar` as the replacement sbt launcher jar, and IDEA makes use of `--addPluginSbtFile` which coursier's sbt-launcher interprets for itself, doesn't understand it, then quits.  	This is why coursier's sbt-extras [adds a `--`](https://github.com/paulp/sbt-extras/pull/239/files#diff-eadc1bacb17aea28d29fefc19868b5e1R572).

I've not spent time thinking through the best solution yet.  But this is a blocker on being able to use coursier/sbt-launcher in IntellIJ.

**Comment by alexarchambault** [comment]
It's been working fine from IntelliJ on my machine. It seems it relies on `~/.sbt/1.0/plugins/idea.sbt` rather than `--addPluginSbtFile` here.

**Comment by alexarchambault** [comment]
But the launcher could always be tweaked, so that it doesn't accept options by default (and passes them all to sbt).
And a Java property, set from the sbt-extras launcher say, could optionally enable argument parsing.

**Comment by dwijnand** [comment]
I think more recent versions of IntelliJ's Scala Plugin use `--addPluginSbtFile` instead of `idea.sbt`?  Perhaps we should tweak the launcher to just pass through and sbt-extras should add something if `coursier_args` isn't empty.
